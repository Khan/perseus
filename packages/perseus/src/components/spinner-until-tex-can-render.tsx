import {CircularSpinner} from "@khanacademy/wonder-blocks-progress-spinner";
import * as React from "react";
import {useState} from "react";

import {getDependencies} from "../dependencies";

type Props = {
    children: () => React.ReactElement;
};

// To minimize layout shift, we sometimes want to display a spinner until
// our math renderer is ready to render math. To do this, we:
// - render a dummy TeX component to force the math renderer to load
// - display a spinner until the TeX component calls its onRender
//   callback, signifying that the math is rendered (from which we can
//   infer that the math renderer has loaded, and is ready to render the
//   actual math on the page).
//
// If we didn't do this, the user might see a blank widget on first render,
// and then the math would pop in a few moments later once the rendering
// library loaded.
export function SpinnerUntilTexCanRender({
    children,
}: Props): React.ReactElement {
    const [dummyTexRendered, setDummyTexRendered] = useState(false);

    if (!dummyTexRendered) {
        const {TeX} = getDependencies();
        const dummyTex = (
            <div style={{display: "none"}}>
                <TeX onRender={() => setDummyTexRendered(true)}>1</TeX>
            </div>
        );
        return (
            <>
                <CircularSpinner />
                {dummyTex}
            </>
        );
    }

    return children();
}
