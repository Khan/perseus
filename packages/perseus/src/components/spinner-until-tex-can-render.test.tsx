import {render, act} from "@testing-library/react";
import * as React from "react";

import "@testing-library/jest-dom"; // Imports custom matchers
import {Dependencies} from "..";
import {testDependencies} from "../../../../testing/test-dependencies";

import {SpinnerUntilTexCanRender} from "./spinner-until-tex-can-render";

describe("SpinnerUntilTexCanRender", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("renders a spinner while waiting for the TeX renderer to load", () => {
        let simulateTeXRendering: undefined | (() => unknown);

        type FakeTeXProps = {children: string; onRender?: () => unknown};
        function FakeTeX({children, onRender}: FakeTeXProps) {
            simulateTeXRendering = onRender;
            return <div className="fake-tex">{children}</div>;
        }
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue({
            ...testDependencies,
            TeX: FakeTeX,
        });

        // Act
        const {container} = render(
            <SpinnerUntilTexCanRender children={() => <p>rendered</p>} />,
        );

        // Assert
        expect(container).toMatchSnapshot("first render: displays a spinner");

        // Act
        act(() => {
            simulateTeXRendering?.();
        });

        // Assert
        expect(container).toMatchSnapshot(
            "second render: displays the sortable",
        );
    });
});
