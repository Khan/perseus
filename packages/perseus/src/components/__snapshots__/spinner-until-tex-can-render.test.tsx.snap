// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinnerUntilTexCanRender renders a spinner while waiting for the TeX renderer to load: first render: displays a spinner 1`] = `
<div>
  <div
    class="default_xu2jcg-o_O-spinnerContainer_agrn11"
  >
    <svg
      height="96"
      viewBox="0 0 96 96"
      width="96"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class="loadingSpinner_1dvgeb3-o_O-inlineStyles_1kzk0y9"
        d="M88.38 46.91a3.818 3.818 0 1 1 7.602 0h.006c.008.362.012.725.012 1.09 0 26.51-21.49 48-48 48S0 74.51 0 48 21.49 0 48 0c.365 0 .728.004 1.09.012v.005a3.818 3.818 0 1 1 0 7.602v.032c-.362-.01-.725-.015-1.09-.015C25.708 7.636 7.636 25.708 7.636 48c0 22.292 18.072 40.364 40.364 40.364 22.292 0 40.364-18.072 40.364-40.364 0-.365-.005-.728-.015-1.09h.032z"
        fill-rule="nonzero"
      />
    </svg>
  </div>
  <div
    style="display: none;"
  >
    <div
      class="fake-tex"
    >
      1
    </div>
  </div>
</div>
`;

exports[`SpinnerUntilTexCanRender renders a spinner while waiting for the TeX renderer to load: second render: displays the sortable 1`] = `
<div>
  <p>
    rendered
  </p>
</div>
`;
